docs/design/mono/components.md

```diff
@@ -0,0 +1,47 @@
+# MonoVM Runtime Components
+
+## Summary
+
+MonoVM runtime components are units of optional functionality that may be provided by the runtime to some workloads.
+
+The idea is to provide some components as optional units that can be loaded dynamically (on workloads that support
+dynamic loading) or that are statically linked into the final application (on workloads that support only static
+linking).
+
+To that end this document describes a methodology for defining and implementing components and for calling component
+functionality from the runtime.
+
+## High level overview
+
+Each component is defined in `src/mono/mono/components`.
+
+The runtime is compiled for different configurations with either static or dynamic linking of components.
+
+When the components are dynamically linked, each component produces a shared library `mono-component-*component-name*`.
+When the components are statically linked, each component produces a static library.
+
+The choice of dynamic or static linking is a global compile-time configuration parameter of the runtime: either all
+components are dynamic or they're all static. In either case, each component may be either present or stubbed out at execution time.
+
+With dynamic linking, all the stubs are built into the runtime, and the runtime probes for the dynamic library of each
+component to see if it is present, or else it calls the stub component.
+
+With static linking, all the present components are statically linked with the runtime into the final app.
+
+Each component exposes a table of functions to the runtime.  The stubs also implement the same table.  In the places
```

Comment:
This is similar to the [local GC](https://github.com/dotnet/runtime/issues/8061) work done several years ago and something I'd like to do with interop in some capacity.  Is there a reason this isn't a general .NET runtime design that both implementations can starting working towards?

Reply:
It depends on the goals, what scenarios does it make better. It would be good to have the goals and scenarios written down in this doc.  

The primary goal for local GC was to make it possible for an app to bundle a GC with bug fix or performance improvement, without affecting global state of the machine. It was addressing a real challenge we had with iterating on GC fixes and improvements with large 1st party customers - the 1st party customers were very hesitant to deploy a .NET Framework patch to test it on a production machines (and these patches were very non-trivial to produce).

Local GC is less valuable now than it was in .NET Framework word. .NET Core supports allows bundling the whole runtime with the app that addresses the primary scenario as well. If we were to do the local GC again today, there is a good chance it would be way below the cut line and it would never get funded.